Mathlib/Combinatorics/Enumerative/Partition/GenFun.lean

```diff
@@ -0,0 +1,227 @@
+    `#(Nat.Partition.odds n)`.
+(TODO: prove these)
+
+The definition of `Nat.Partition.genFun` ignores the value of $f(0, c)$ and $f(i, 0)$. The formula
+can be interpreted as assuming $f(i, 0) = 1$ and $f(0, c) = 0$ for $c \ne 0$. In theory we could
+respect the actual value of $f(0, c)$ and $f(i, 0)$, but it makes the otherwise finite sum and
+product potentially infinite.
+-/
+
+open PowerSeries
+open scoped PowerSeries.WithPiTopology
+
+namespace Nat.Partition
+variable {M : Type*} [CommSemiring M]
+
+/-- Generating function associated with character $f(i, c)$ for partition functions, where $i$ is a
+part of the partition, and $c$ is the count of that part in the partition. The character function is
+multiplied within one `n.Partition`, and summed among all `n.Partition` for a fixed `n`. This way,
+each `n` is assigned a value, which we use as the coefficients of the power series. -/
+def genFun (f : ℕ → ℕ → M) : M⟦X⟧ :=
+  PowerSeries.mk fun n ↦ ∑ p : n.Partition, ∏ i ∈ p.parts.toFinset, f i (p.parts.count i)
+
+variable [TopologicalSpace M]
+
+/-- The infinite sum in the formula `Nat.Partition.hasProd_genFun` always converges. -/
+theorem summable_genFun_term (f : ℕ → ℕ → M) (i : ℕ) :
+    Summable fun j ↦ f (i + 1) (j + 1) • (X : M⟦X⟧) ^ ((i + 1) * (j + 1)) := by
+  rcases subsingleton_or_nontrivial M with h | h
+  · simpa [Subsingleton.eq_zero] using summable_zero
+  apply WithPiTopology.summable_of_tendsto_order_atTop_nhds_top
+  refine ENat.tendsto_nhds_top_iff_natCast_lt.mpr (fun n ↦ Filter.eventually_atTop.mpr ⟨n, ?_⟩)
+  intro m hm
+  grw [PowerSeries.smul_eq_C_mul, ← le_order_mul]
+  refine lt_add_of_nonneg_of_lt (by simp) ?_
+  rw [order_X_pow]
+  norm_cast
+  grind
+
+variable [T2Space M]
+
+private theorem aux_injOn_sub_one_parts {n : ℕ} (p : Partition n) :
+    Set.InjOn (fun x ↦ x - 1) p.parts.toFinset := by
+  intro a ha b hb
+  exact tsub_inj_left (p.parts_pos (by simpa using ha)) (p.parts_pos (by simpa using hb))
+
+private theorem aux_map_sub_one_parts_subset {n : ℕ} (p : Partition n) :
+    (Multiset.map (fun x ↦ x - 1) p.parts).toFinset ⊆ Finset.range n := by
+  intro m
+  rw [Multiset.mem_toFinset, Finset.mem_range]
+  suffices ∀ x ∈ p.parts, x - 1 = m → m < n by simpa
+  rintro x h1 rfl
+  exact Nat.sub_one_lt_of_le (p.parts_pos h1) (le_of_mem_parts h1)
+
+private theorem aux_mapsTo_sub_one_parts {n : ℕ} (p : Partition n) :
+    Set.MapsTo (fun x ↦ x - 1) p.parts.toFinset (Finset.range n) := by
+  intro a ha
+  apply Finset.mem_of_subset (aux_map_sub_one_parts_subset p)
+  rw [Multiset.mem_toFinset, Multiset.mem_map]
+  exact ⟨a, by simpa using ha⟩
+
+private def toFinsuppAntidiag {n : ℕ} (p : Partition n) : ℕ →₀ ℕ where
+  toFun m := p.parts.count (m + 1) * (m + 1)
+  support := (p.parts.map (· - 1)).toFinset
+  mem_support_toFun m := by
+    suffices (∃ a ∈ p.parts, a - 1 = m) ↔ m + 1 ∈ p.parts by simpa
+    grind [→ Partition.parts_pos]
```

Comment:
I think this function is generally interesting. Could you make it public and move it to an earlier file?

Reply:
It's because the index in hasProd_genFun is shifted, and that's because PNat is hard to use with infinite sum / product (I attempted a version of that and it resulted in longer and messy proof)

Reply:
Ah then it's pretty clear to me that the public version of the function should not shifted, and that you should then compose with index shifting within your proof